diff --git a/Configure b/Configure
index b286dd0678..853e8853e1 100755
--- a/Configure
+++ b/Configure
@@ -327,7 +327,7 @@ $config{sdirs} = [
     "bn", "ec", "rsa", "dsa", "dh", "sm2", "dso", "engine",
     "buffer", "bio", "stack", "lhash", "rand", "err",
     "evp", "asn1", "pem", "x509", "x509v3", "conf", "txt_db", "pkcs7", "pkcs12", "comp", "ocsp", "ui",
-    "cms", "ts", "srp", "cmac", "ct", "async", "kdf", "store"
+    "cms", "ts", "srp", "cmac", "ct", "async", "kdf", "store", "oqs", "composite"
     ];
 # test/ subdirectories to build
 $config{tdirs} = [ "ossl_shim" ];
@@ -359,6 +359,7 @@ my @disablables = (
     "cmac",
     "cms",
     "comp",
+    "composite",
     "crypto-mdebug",
     "crypto-mdebug-backtrace",
     "ct",
diff --git a/crypto/asn1/standard_methods.h b/crypto/asn1/standard_methods.h
index 9e29c85e36..81e6b01a37 100644
--- a/crypto/asn1/standard_methods.h
+++ b/crypto/asn1/standard_methods.h
@@ -29,6 +29,9 @@ static const EVP_PKEY_ASN1_METHOD *standard_methods[] = {
     &dsa_asn1_meths[3],
     &dsa_asn1_meths[4],
 #endif
+#ifndef OPENSSL_NO_COMPOSITE
+    &composite_asn1_meth,
+#endif
 #ifndef OPENSSL_NO_EC
     &eckey_asn1_meth,
 #endif
diff --git a/crypto/ec/build.info b/crypto/ec/build.info
index 62d81cb7bb..a1e673e347 100644
--- a/crypto/ec/build.info
+++ b/crypto/ec/build.info
@@ -8,7 +8,6 @@ SOURCE[../../libcrypto]=\
         ecdsa_ossl.c ecdsa_sign.c ecdsa_vrf.c curve25519.c ecx_meth.c \
         curve448/arch_32/f_impl.c curve448/f_generic.c curve448/scalar.c \
         curve448/curve448_tables.c curve448/eddsa.c curve448/curve448.c \
-	oqs_meth.c \
         {- $target{ec_asm_src} -}
 
 GENERATE[ecp_nistz256-x86.s]=asm/ecp_nistz256-x86.pl \
diff --git a/crypto/evp/pmeth_lib.c b/crypto/evp/pmeth_lib.c
index 09f7622daa..39daca66f5 100644
--- a/crypto/evp/pmeth_lib.c
+++ b/crypto/evp/pmeth_lib.c
@@ -112,6 +112,11 @@ static const EVP_PKEY_METHOD *standard_methods[] = {
     &p256_sphincsshake256128frobust_pkey_meth,
     &rsa3072_sphincsshake256128frobust_pkey_meth,
 ///// OQS_TEMPLATE_FRAGMENT_LIST_PKEY_METHS_END
+
+#ifndef OPENSSL_NO_COMPOSITE
+    &composite_pkey_meth,
+#endif
+
 };
 
 DECLARE_OBJ_BSEARCH_CMP_FN(const EVP_PKEY_METHOD *, const EVP_PKEY_METHOD *,
diff --git a/crypto/objects/obj_xref.txt b/crypto/objects/obj_xref.txt
index ad32d3463f..68809d6408 100644
--- a/crypto/objects/obj_xref.txt
+++ b/crypto/objects/obj_xref.txt
@@ -68,6 +68,8 @@ p256_sphincsshake256128frobust	sha512	p256_sphincsshake256128frobust
 rsa3072_sphincsshake256128frobust	sha512	rsa3072_sphincsshake256128frobust
 ##### OQS_TEMPLATE_FRAGMENT_LIST_OQS_SIGS_END
 
+composite       undef   composite
+composite_alt   undef   composite_alt
 
 # Alternative deprecated OIDs. By using the older "rsa" OID this
 # type will be recognized by not normally used.
diff --git a/crypto/objects/objects.txt b/crypto/objects/objects.txt
index efc1d7962e..e316045d87 100644
--- a/crypto/objects/objects.txt
+++ b/crypto/objects/objects.txt
@@ -1676,6 +1676,21 @@ dstu4145le 2 7 : uacurve7 : DSTU curve 7
 dstu4145le 2 8 : uacurve8 : DSTU curve 8
 dstu4145le 2 9 : uacurve9 : DSTU curve 9
 
+# =========================================
+# Composite Crypto and Combined Crypto OIDs
+# =========================================
+#
+# Composite Crypto OID:
+# 1.3.6.1.4.1.18227.2.1
+#
+# CompositeOR Crypto OID:
+# 1.3.6.1.4.1.18227.2.2
+#
+
+1 3 6 1 4 1 18227 : id-pen-openca : OpenCA Labs Private Enterprise Name
+id-pen-openca 2 1 : composite     : Composite Crypto With Combined Keys
+id-pen-openca 2 2 : composite_alt : Composite Crypto With Alternative Keys
+
 ##### OQS_TEMPLATE_FRAGMENT_LIST_KEMS_START
  : oqs_kem_default : oqs_kem_default
  : p256_oqs_kem_default : p256_oqs_kem_default
diff --git a/crypto/x509/x509type.c b/crypto/x509/x509type.c
index 1e7525d270..6c3c1554df 100644
--- a/crypto/x509/x509type.c
+++ b/crypto/x509/x509type.c
@@ -101,6 +101,10 @@ int X509_certificate_type(const X509 *x, const EVP_PKEY *pkey)
     case NID_id_GostR3410_2012_512:
         ret = EVP_PKT_EXCH | EVP_PKT_SIGN;
         break;
+    case NID_composite:
+    case NID_composite_alt:
+        ret = EVP_PKT_SIGN;
+        break;
     default:
         break;
     }
diff --git a/include/crypto/asn1.h b/include/crypto/asn1.h
index 40896eee39..700e0edd99 100644
--- a/include/crypto/asn1.h
+++ b/include/crypto/asn1.h
@@ -130,6 +130,10 @@ extern const EVP_PKEY_ASN1_METHOD sphincsshake256128frobust_asn1_meth;
 extern const EVP_PKEY_ASN1_METHOD p256_sphincsshake256128frobust_asn1_meth;
 extern const EVP_PKEY_ASN1_METHOD rsa3072_sphincsshake256128frobust_asn1_meth;///// OQS_TEMPLATE_FRAGMENT_DEFINE_ASN1_METHS_END
 
+#ifndef OPENSSL_NO_COMPOSITE
+extern const EVP_PKEY_ASN1_METHOD composite_asn1_meth;
+#endif
+
 /*
  * These are used internally in the ASN1_OBJECT to keep track of whether the
  * names and data need to be free()ed
diff --git a/include/crypto/evp.h b/include/crypto/evp.h
index 057e8278ba..26a7f9a0b0 100644
--- a/include/crypto/evp.h
+++ b/include/crypto/evp.h
@@ -156,6 +156,10 @@ extern const EVP_PKEY_METHOD sphincsshake256128frobust_pkey_meth;
 extern const EVP_PKEY_METHOD p256_sphincsshake256128frobust_pkey_meth;
 extern const EVP_PKEY_METHOD rsa3072_sphincsshake256128frobust_pkey_meth;///// OQS_TEMPLATE_FRAGMENT_DEFINE_EVP_METHS_END
 
+#ifndef OPENSSL_NO_COMPOSITE
+extern const EVP_PKEY_ASN1_METHOD composite_asn1_meth;
+#endif
+
 struct evp_md_st {
     int type;
     int pkey_type;
@@ -342,6 +346,12 @@ BLOCK_CIPHER_def_ofb(cname, kstruct, nid, key_len, iv_len, cbits, \
 BLOCK_CIPHER_def_ecb(cname, kstruct, nid, block_size, key_len, flags, \
                      init_key, cleanup, set_asn1, get_asn1, ctrl)
 
+
+# ifndef OPENSSL_NO_COMPOSITE
+#  define EVP_PKEY_assign_COMPOSITE(pkey,comp_key) EVP_PKEY_assign((pkey),EVP_PKEY_COMPOSITE,\
+                                        (char *)(comp_key))
+# endif
+
 /*-
 #define BLOCK_CIPHER_defs(cname, kstruct, \
                                 nid, block_size, key_len, iv_len, flags,\
diff --git a/include/openssl/evp.h b/include/openssl/evp.h
index 5375528c1b..768b1fdf97 100644
--- a/include/openssl/evp.h
+++ b/include/openssl/evp.h
@@ -63,6 +63,21 @@
 # define EVP_PKEY_ED25519 NID_ED25519
 # define EVP_PKEY_X448 NID_X448
 # define EVP_PKEY_ED448 NID_ED448
+
+// ========================
+// Composite Crypto Support
+// ========================
+
+#ifndef OPENSSL_NO_COMPOSITE
+# define EVP_PKEY_COMPOSITE NID_composite
+# define EVP_PKEY_COMPOSITE_OR NID_composite_alt
+#endif
+
+# define EVP_PKEY_CTRL_COMPOSITE_PUSH    0x201
+# define EVP_PKEY_CTRL_COMPOSITE_POP     0x202
+# define EVP_PKEY_CTRL_COMPOSITE_DEL     0x204
+# define EVP_PKEY_CTRL_COMPOSITE_CLEAR   0x205
+
 ///// OQS_TEMPLATE_FRAGMENT_DEFINE_EVP_PKEYS_START
 # define EVP_PKEY_OQS_SIG_DEFAULT NID_oqs_sig_default
 # define EVP_PKEY_P256_OQS_SIG_DEFAULT NID_p256_oqs_sig_default
